R/variance_ratio.R

```diff
@@ -33,100 +33,162 @@
+#'
 #'  #Calculate the variance ratio and CIs for each replicate
-#'  
-#'  res_withinreplicates <- variance_ratio(knz_001d, time.var = "year", species.var = "species", 
+#'
+#'  res_withinreplicates <- variance_ratio(knz_001d, time.var = "year", species.var = "species",
 #'  abundance.var = "abundance", bootnumber = 1, replicate = "subplot", average.replicates = FALSE)
-variance_ratio <- function(df, time.var, species.var, abundance.var, bootnumber, replicate.var=NA,
-                        li=0.025, ui=0.975,  average.replicates=TRUE) {
-  
+variance_ratio <- function(df, time.var, species.var, abundance.var,
+                           bootnumber, replicate.var,
+                           average.replicates = TRUE, level = 0.95, ...) {
+
   # check to make sure abundance is numeric data
   check_numeric(df, time.var, abundance.var)
-   
+
   # if no replicates, calculate single variance ratio
-   if(is.na(replicate.var)) {
-        check_single_onerep(df, time.var, species.var)
-        VR <- variance_ratio_longformdata(df, time.var, species.var, abundance.var)
+  if  (missing(replicate.var)) {
+
+    check_single_onerep(df, time.var, species.var)
+    VR <- variance_ratio_longformdata(df, time.var, species.var, abundance.var)
+
+    ## null models
+
+    nullval <- cyclic_shift(df, time.var = time.var,
+                            species.var = species.var,
+                            abundance.var = abundance.var,
+                            method = variance_ratio_matrixdata,
+                            bootnumber = bootnumber)
+
+    nullout <- confint(nullval)
+
+    # bind actual value and CI
+    output <- cbind(nullout, VR)
+
+  } else {
+
+    # if multiple replicates, check all replicates have values
+    check_single(df, time.var, species.var, replicate.var)
+    ## if you give a replicate, it must be a factor. This gives users responsibility for order.
+    assertthat::assert_that(is.factor(df[[replicate.var]]))
+
+
+    # calculate average variance ratio across replicates
+    if (average.replicates == TRUE) {
+      ## check the data
+      check_multispp(df, species.var, replicate.var)
+
+      # split the dataset into replicates, according to replicate.var
+      df <- df[order(df[[replicate.var]]),]
+      X <- split(df, df[replicate.var])
+
+      VR <- mean(unlist(lapply(X, FUN = variance_ratio_longformdata, time.var, species.var, abundance.var)))
+
+      ##
+      null_list <- lapply(X, FUN = cyclic_shift,  time.var = time.var,
+                          species.var = species.var,
+                          abundance.var = abundance.var,
+                          method = variance_ratio_matrixdata,
+                          bootnumber = bootnumber)
+
+      ## confidence intervals
+      null_output <- lapply(null_list, `[[`, i = "out")
+
+      null_means <- vector(mode = "numeric", length = bootnumber)
+
+      for (B in 1:bootnumber) {
+        null_means[B] <- mean(vapply(null_output, `[[`, 1, i = B))
+      }
```

Comment:
@gavinsimpson @aammd Currently we cast a new species x time matrix for each replicate, and don't include species that are never present in the replicate. We did this partly to reduce data frame size, and partly to encourage tidy input data/model how to manipulate it. Some of the syntheses groups that we worked with to test the package had huge species x (site, time) matrices with most species in any given site being all 0s - this was frustrating and super hard to QA, so we opted to require the input data be formatted differently. 

Logistically it would require some overhaul, but we could shift to casting with all species in the overall dataset (thus, all replicates having the same species, that are sometimes all 0), then applying the same random starting points for each species across replicates. To do this I think we'd go back to appending the dataframe to itself, and saving/applying starting points - could you do this with shuffleSeries? Conceptually, though, I think it's pretty standard to have each replicate re-randomized to increase permutation number. You two okay with keeping it that way, and clarifying the help doc and MS text to spell out that that is what it is doing?